connections/stream-migration.md

```diff
@@ -0,0 +1,217 @@
+The protocol id should be `/libp2p/stream-migration`.
+
+### Stream IDs
+
+Stream IDs are a uint64. They are defined by the stream initiator and conveyed to
+the responder in the `StreamMigration` message. The ID should be unique to the
+nodes involved. Even across connections. In other words, for two peers A and B,
+every labelled stream between them should have a unique uint64 ID. To ensure
+that this is true across implementations IDs generated by the lower peer id
+should be even, and IDs generated by the higher peer ID should be odd.
+
+Here is a possible strategy that implementors could use when labelling a
+stream, but is not required as long as the above invariants hold true.
+ 1. Define an atomic uint64 as a counter.
+ 1. Grab and increment the counter, call this `ID`.
+ 1. Bitshift-left the `ID` by one (i.e. multiply it by two).
+ 1. Check if we are the lower peer ID.
+    1. If yes, do nothing
+    1. If no, add one to the `ID`
+ 1. Use `ID` as the ID for the stream.
+
+Note that this reduces the total counter space to be 63bits since the lowest bit
+is used to signal which node labelled this stream. This limit should be plenty
+high in practice, as no node will have more than `2^63` streams.
+
+### Stream Migration Messages
+
+Messages for stream migration are Protobuf messages defined in
+[./stream-migration/streammigration.proto](./stream-migration/streammigration.proto).
+
+The first StreamMigration message sent over the wire by the stream initiator can
+be one of:
+1. Label. This labels this stream with an ID (defined above).
+1. Migrate. This starts the migration from a given id to this stream, it also
+   labels the stream with an ID.
+
+The responder should only respond to the `Migrate` message with a `AckMigrate`
+message that may optionally deny the migration.
+
+## Who moves the stream
+
+This protocol makes no assumption on which node starts the migration. Either
+node may start the migration. However extensions to this protocol may want to
+designate a single node to start the migration (i.e. picking the node with the
+lowest peer id).
+
+## Picking the best connection
+
+Picking the best connection to migrate streams towards is outside the scope of
+this protocol. However, it's important to note that only the initiator of the
+stream migration is concerned with which connection to pick. Both sides do not
+have to have the same deterministic notion of what is best.
+
+### Resets
+
+If either stream is "reset" before both ends are closed, both streams must be
+reset and the stream as a whole should be considered "aborted" (reset).
+
+### Half closed streams
+
+The final migrated stream should look the same as the initial stream. If the
+initial stream `1` was half closed, then the final migrated stream `2` should
+also be half closed. Note this may involve an extra step by one of the nodes.
+If a node, had closed writes to its old stream before migration it should also
+close writes to the new stream after migration.
```

Comment:
We should probably call out what happens if a node sends new data if the stream was already closed in that direction: that is a connection error.